### ISSUE TITLE
Getting error "Sizes of columns doesn't match" when reading json from kafka table engine 

### ISSUE THREAD

**Issue by brefre** [ClickHouse/ClickHouse#8003]
**Describe the bug or unexpected behaviour**
After upgrading from 19.16.4.12 to 19.17.4.11, selecting data from kafka table engine gives this error in the error log:

2019.12.03 06:11:00.367034 [ 27 ] {} <Error> void DB::StorageKafka::threadFunc(): Code: 9, e.displayText() = DB::Exception: Sizes of columns doesn't match: date: 9001, _timestamp: 0, Stack trace

**How to reproduce**

My topic has json-objects with the field date encoded as a string, ie 

```
{ "date": "2019-01-01", "foo1": "bar", "foo2": "bar" }
```

```
CREATE TABLE kafka.tmp_test
(
    `date` Date
)
ENGINE = Kafka('xxx:9092', 'xx', 'xxgroup, 'JSONEachRow')

Ok.

0 rows in set. Elapsed: 0.002 sec.

db :) select * from tmp_test limit 1;

SELECT *
FROM tmp_test
LIMIT 1

Ok.

0 rows in set. Elapsed: 1.003 sec.

db :)
```

**Error message and/or stacktrace**
``` 
2019.12.03 06:13:10.782027 [ 21 ] {} <Error> void DB::StorageKafka::threadFunc(): Code: 9, e.displayText() = DB::Exception: Sizes of columns doesn't match: date: 20831, _timestamp: 0, Stack trace:

0. 0x3512b60 StackTrace::StackTrace() /usr/bin/clickhouse
1. 0x351cdaf DB::Exception::Exception(std::string const&, int) /usr/bin/clickhouse
2. 0x60141cf DB::Block::checkNumberOfRows(bool) const /usr/bin/clickhouse
3. 0x69dc27e DB::KafkaBlockInputStream::readImpl() /usr/bin/clickhouse
4. 0x6073008 DB::IBlockInputStream::read() /usr/bin/clickhouse
5. 0x609cbe6 DB::copyData(DB::IBlockInputStream&, DB::IBlockOutputStream&, std::atomic<bool>*) /usr/bin/clickhouse
6. 0x66d7206 DB::StorageKafka::streamToViews() /usr/bin/clickhouse
7. 0x66d7770 DB::StorageKafka::threadFunc() /usr/bin/clickhouse
8. 0x66ed3c4 DB::BackgroundSchedulePoolTaskInfo::execute() /usr/bin/clickhouse
9. 0x66edf92 DB::BackgroundSchedulePool::threadFunction() /usr/bin/clickhouse
10. 0x66edff4 ? /usr/bin/clickhouse
11. 0x3554f13 ThreadPoolImpl<std::thread>::worker(std::_List_iterator<std::thread>) /usr/bin/clickhouse
12. 0x791d69f ? /usr/bin/clickhouse
13. 0x7fc42d6b9dc5 start_thread /usr/lib64/libpthread-2.17.so
14. 0x7fc42d1e476d clone /usr/lib64/libc-2.17.so
 (version 19.17.4.11)
2019.12.03 06:13:17.858559 [ 25 ] {} <Error> void DB::StorageKafka::threadFunc(): Code: 9, e.displayText() = DB::Exception: Sizes of columns doesn't match: date: 3466, _timestamp: 0, Stack trace:

0. 0x3512b60 StackTrace::StackTrace() /usr/bin/clickhouse
1. 0x351cdaf DB::Exception::Exception(std::string const&, int) /usr/bin/clickhouse
2. 0x60141cf DB::Block::checkNumberOfRows(bool) const /usr/bin/clickhouse
3. 0x69dc27e DB::KafkaBlockInputStream::readImpl() /usr/bin/clickhouse
4. 0x6073008 DB::IBlockInputStream::read() /usr/bin/clickhouse
5. 0x609cbe6 DB::copyData(DB::IBlockInputStream&, DB::IBlockOutputStream&, std::atomic<bool>*) /usr/bin/clickhouse
6. 0x66d7206 DB::StorageKafka::streamToViews() /usr/bin/clickhouse
7. 0x66d7770 DB::StorageKafka::threadFunc() /usr/bin/clickhouse
8. 0x66ed3c4 DB::BackgroundSchedulePoolTaskInfo::execute() /usr/bin/clickhouse
9. 0x66edf92 DB::BackgroundSchedulePool::threadFunction() /usr/bin/clickhouse
10. 0x66edff4 ? /usr/bin/clickhouse
11. 0x3554f13 ThreadPoolImpl<std::thread>::worker(std::_List_iterator<std::thread>) /usr/bin/clickhouse
12. 0x791d69f ? /usr/bin/clickhouse
13. 0x7fc42d6b9dc5 start_thread /usr/lib64/libpthread-2.17.so
14. 0x7fc42d1e476d clone /usr/lib64/libc-2.17.so
 (version 19.17.4.11)
```

**Comment by filimonov** [comment]
Thanks for the report i think i can fix that, right now working with code around that place. Just to ensure - your messages in kafka have no timestamps?

**Comment by brefre** [comment]
Hi,

This is a good question. I assume you are talking about the internal timestamps in the kafka messages, and not my JSON structure. 

Many months ago, we upgraded our kafka cluster from a older version. I have found that topics created in the older version of kafka do not have timestamps, but newly created topics do have timestamps, so currently its a mix. I am not 100% certain about this topic, but I believe this is an old topic, and because of this, it does not have timestamps.

Thank you for contributing to great open source software.

**Comment by niting112** [comment]
@brefre @filimonov Is this a bug for some particular use case or in general when using kafka engine ?

**Comment by filimonov** [comment]
> @filimonov Is this a bug for some particular use case or in general when using kafka engine ?

1) it is already fixed, but fix was not released yet.
2) you hit that bug if your messages in Kafka have no timestamp, or if you use Native format.

> Prevent message duplication when producing Kafka table has any MVs selecting from it).

I think 'prevent duplication' misleads you. Do you really insert into the same Kafka table where some MV attached?

**Comment by filimonov** [comment]
Fixed in #8016 
Released in 19.16.10 and newer
